stores/filestore.go

```diff
@@ -3208,6 +3241,148 @@ func (ms *FileMsgStore) lookup(seq uint64) (*pb.MsgProto, error) {
+		crc := util.ByteOrder.Uint32(header[4:recordHeaderSize])
+		// check CRC against what was stored
+		if c := crc32.Checksum(buf[recordHeaderSize:totalSize], ms.fstore.crcTable); c != crc {
+			return buf, fmt.Errorf("corrupted data, expected crc to be 0x%08x, got 0x%08x", crc, c)
+		}
+	}
+	return buf, nil
+}
+
+// Read from disk and returns the message for the given `seq` but perform
+// a single disk I/O with large buffer to read possibly several subsequent
+// messages. These are added to the cache which will speed up lookups for
+// next message sequence.
+func (ms *FileMsgStore) readAheadMsgs(fslice *fileSlice, seq uint64) (*pb.MsgProto, error) {
+
+	// Compute the offset in the index file itself.
+	idxFileOffset := 4 + (int64(seq-fslice.firstSeq)+int64(fslice.rmCount))*msgIndexRecSize
+	// Then position the file pointer of the index file.
+	if _, err := fslice.idxFile.handle.Seek(idxFileOffset, io.SeekStart); err != nil {
+		return nil, err
+	}
+
+	// Keep track of individual size of messages we are going to bulk read
+	var _msgSizes [2048]int
+	msgSizes := _msgSizes[:0]
+
+	// Keep track of total number of messages and size
+	totalMsgs := 0
+	totalMsgsSize := 0
+	firstMsgOffset := int64(0)
+
+	// Indexes are fixed size. Read a page (4K) worth of them.
+	// Loop if needed if total message size is below readBufSize.
+	idxsBuf := [128 * msgIndexRecSize]byte{}
+	idxSeq := seq
+	maxIdxs := int(fslice.lastSeq - seq + 1)
+	done := false
+	idxBufSize := cap(idxsBuf)
+	for iter := 0; !done; iter++ {
+		if (iter+1)*128 > maxIdxs {
+			idxBufSize = (maxIdxs - (iter * 128)) * msgIndexRecSize
+			// Will be done after reading this page of indexes
+			done = true
+		}
+		b, err := io.ReadFull(fslice.idxFile.handle, idxsBuf[:idxBufSize])
+		if err != nil {
+			return nil, err
+		}
+		roffset := 0
+		numIdxs := b / msgIndexRecSize
+		for i := 0; i < numIdxs; i++ {
+			seqInIndexFile, msgIndex, err := ms.readIndexFromBuffer(idxsBuf[roffset:])
+			if err != nil {
+				return nil, err
+			}
+			if seqInIndexFile != idxSeq {
+				return nil, fmt.Errorf("wrong sequence, wanted %v got %v", idxSeq, seqInIndexFile)
+			}
+			if msgIndex == nil {
+				return nil, nil
+			}
+			recSize := int(msgIndex.msgSize + recordHeaderSize)
+			// Stop when we are over the limit, but we need to read at least one message
+			if (iter > 0 || i > 0) && totalMsgsSize+recSize > ms.readBufSize {
+				done = true
+				break
+			}
+			if iter == 0 && i == 0 {
+				firstMsgOffset = msgIndex.offset
+			}
+			totalMsgs++
+			totalMsgsSize += recSize
+			msgSizes = append(msgSizes, int(msgIndex.msgSize))
+			roffset += msgIndexRecSize
+			idxSeq++
+		}
+	}
+
+	// Position the data file to the offset of the first message we need to read.
+	file := fslice.file.handle
+	if _, err := file.Seek(firstMsgOffset, io.SeekStart); err != nil {
+		return nil, err
+	}
+	// Use a buffer that can hold all the message(s)
+	ms.tmpMsgBuf = util.EnsureBufBigEnough(ms.tmpMsgBuf, totalMsgsSize)
```

Comment:
Do we reuse the via sync.Pool? Should we? Them and the index structs maybe?

Reply:
So ms.tmpMsgBuf is a FileMsgStore member and was used so far for reading/writing single messages.
I will probably replace with dedicated buffer (I have one for write calls) or as you say use a sync.Pool, which may be better/simpler than the implementation I had for shrinking the write buffer...

Reply:
Sounds good.